### PR TITLE
adding comments for sax parser

### PR DESCRIPTION
**Pull request by galsalomon66**
Signed-off-by: gal salomon <gal.salomon@gmail.com>

### CODE REVIEW

File: json_parser.cpp [hunk 1]
```diff
@@ -67,148 +67,156 @@ class Valuesax {
...
+  public:
     std::vector < std::pair < std::string, Valuesax>>  mymap;
     std::vector <char> vect;
     std::string keyname{"first"};
     std::string keyvalue;
     bool valuep{false};
     int start_counter{0};
     std::vector<std::string> mystack;
-    std::unordered_map<int, std::string> key_stack;
+    std::unordered_map<int, std::string> key_stack;//TODO why not using an array?
```
Comment (galsalomon66) [inline json_parser.cpp:118]: why not using an array?

File: json_parser.cpp [hunk 2]
```diff
@@ -67,148 +67,156 @@ class Valuesax {
...
       valuep = true; 
       return true; }
+
     bool Bool(bool b) {
       mymap.push_back(std::make_pair(keyvalue, Valuesax::Parse(b)));
       valuep = true;
       return true; }
-    bool Int(int i) { 
+
+    bool Int(int i) {//TODO to complete
```
Comment (galsalomon66) [inline json_parser.cpp:130]: to complete

File: json_parser.cpp [hunk 3]
```diff
@@ -67,148 +67,156 @@ class Valuesax {
...
+    bool Int(int i) {//TODO to complete 
       return true; }
+
     bool Uint(unsigned u) {
       mymap.push_back(std::make_pair(keyvalue, Valuesax::Parse(u)));
       valuep = true;
       return true; }
-    bool Int64(int64_t i) { 
+
+    bool Int64(int64_t i) {//TODO to complete
```
Comment (galsalomon66) [inline json_parser.cpp:138]: to complete

File: json_parser.cpp [hunk 4]
```diff
@@ -67,148 +67,156 @@ class Valuesax {
...
       mymap.push_back(std::make_pair(keyvalue, Valuesax::Parse(u)));
       valuep = true;
       return true; }
-    bool Int64(int64_t i) { 
+
+    bool Int64(int64_t i) {//TODO to complete  
       return true; }
-    bool Uint64(uint64_t u) { 
+
+    bool Uint64(uint64_t u) {//TODO to complete
```
Comment (galsalomon66) [inline json_parser.cpp:141]: to complete

File: json_parser.cpp [hunk 5]
```diff
@@ -67,148 +67,156 @@ class Valuesax {
...
-        if (keyname != "first") {
-        mystack.push_back(keyname);
-        start_counter = vect.size();
-        key_stack[start_counter] = keyname;
-        }
+
+    bool StartObject() {
+      if(!valuep) {
+	if (mystack.size() == 0 || mystack[mystack.size() - 1] != keyname) {
+	  if (keyname != "first") {//TODO "first" could be real key-name, it could not be used as a state
```
Comment (galsalomon66) [inline json_parser.cpp:158]: "first" could be a real key name, it could not be used as a state

File: json_parser.cpp [hunk 6]
```diff
@@ -67,148 +67,156 @@ class Valuesax {
...
-    
-    for (const auto& i: stack) {
-      keyvalue += i + '/';
-    }
-    keyname = str;
-      
-    return true;
-    }
+    bool Key(const char* str, SizeType length, bool copy) {
+      std::vector<std::string> stack{mystack}; //TODO on each call it construct stack with mystack
```
Comment (galsalomon66) [inline json_parser.cpp:170]: on each call it construct stack with mystack

File: json_parser.cpp [hunk 7]
```diff
@@ -67,148 +67,156 @@ class Valuesax {
...
-  }
-    return true; 
-  }
+      return true; 
+    }
 };
 
 std::string parse_json_sax(const char* file_name)
-{
+{// the parser is actually load the whole input (several GB .... )
```
Comment (galsalomon66) [inline json_parser.cpp:219]: the parser is actually load the whole input (several GB .... )

File: json_parser.cpp [hunk 8]
```diff
@@ -218,7 +226,7 @@ std::string parse_json_sax(const char* file_name)
   if (in.is_open()) {
     std::string line;
     while (getline(in, line)) {
-        stringFromStream.append(line + "\n");
+      stringFromStream.append(line + "\n"); //the entire file is loaded into stringFromStream
```
Comment (galsalomon66) [inline json_parser.cpp:229]: the entire file is loaded into stringFromStream

File: json_parser.cpp [hunk 9]
```diff
@@ -230,7 +238,7 @@ std::string parse_json_sax(const char* file_name)
   StringStream ss(data);
   reader.Parse(ss, handler);
 
-  for (auto const& i: handler.mymap) {
+  for (auto const& i: handler.mymap) { //mymap contains all keys/vause, could be huge amount of data
```
Comment (galsalomon66) [inline json_parser.cpp:241]: mymap contains all keys/value, could be a huge amount of data

File: json_parser.cpp (hunk 1 above)
Comment (albin-antony) [reply json_parser.cpp:118]: Chose unordered_map as keys are not contiguous.

Comment (galsalomon66) [reply json_parser.cpp:118]: the access is random and limited by the depth size of the JSON, quite ideal for stl::vector.
the vector(array) could initialize with some "big" size, let's say 400.
the vector can be resized upon index is too big (it's not happening too much)

Comment (albin-antony) [reply json_parser.cpp:118]: Yes, the access is random. But, assignment in this case, I think the start_counter values are not contiguous

Comment (galsalomon66) [reply json_parser.cpp:118]: indeed, the start_counter values are not contiguous ( 1 --> json-max-depth).
that's why it needs to get initialized at the beginning.

File: sax.h [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+public:
+    enum Type {
+        Decimal,
+        Double,
+        String,
+        Bool,
+        Null
+    };
+
+    static Valuesax Parse(std::string const& s) {
```
Comment (galsalomon66) [inline sax.h:27]: please remove the static for Valuesax::parse 
no need for that

File: test_json.cpp [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+  //read whole file content into allocated buffer
+  std::string file_content(sz, '\0');
+  dom_input_file.read((char*)file_content.data(),sz);
+
+  Document document;
+  document.Parse(file_content.data());
+
+  if (document.HasParseError()) {
+    std::cout<<"parsing error"<< std::endl;
+    return std::string("parsing error");
```
Comment (galsalomon66) [inline test_json.cpp:46]: in case of error, the flow can not proceed.

File: sax.h (hunk 10 above)
Comment (albin-antony) [reply sax.h:27]: But, I am calling parse functions in MyHandler class without instantiating an object of ValueSax

Comment (galsalomon66) [reply sax.h:27]: using static is problematic.
The static function can modify only static data members.
it breaks the object-oriented paradigm.
the static data member is problematic, it's not thread-safe.
thus, we should avoid using static where possible.

there is no need for static in this case.

ValueSax::Parse is actually instantiating ValueSax in the function body, and returning it by value (it means to copy, not efficient)

**Pull request closed without merging**